File: src/algorithms/pathfinding/heuristic-search/weighted-a-star/weighted-a-star.test.ts
import { describe, it, expect } from "vitest";
import type { GridCell } from "@/types";
import { weightedAStar } from "./sources/weighted-a-star.ts?fn";

function createEmptyGrid(rows: number, cols: number): GridCell[][] {
  return Array.from({ length: rows }, (_, rowIndex) =>
    Array.from({ length: cols }, (_, colIndex) => ({
      row: rowIndex,
      col: colIndex,
      type: "empty" as const,
      state: "default" as const,
    })),
  );
}

function setCell(grid: GridCell[][], row: number, col: number, type: GridCell["type"]): void {
  const gridRow = grid[row];
  if (gridRow) {
    const cell = gridRow[col];
    if (cell) cell.type = type;
  }
}

describe("weightedAStar", () => {
  it("finds a path on an empty grid", () => {
    const grid = createEmptyGrid(5, 5);
    setCell(grid, 0, 0, "start");
    setCell(grid, 4, 4, "end");

    const result = weightedAStar(grid, [0, 0], [4, 4], 1.5);

    expect(result.path.length).toBeGreaterThan(0);
    expect(result.path[0]).toEqual([0, 0]);
    expect(result.path[result.path.length - 1]).toEqual([4, 4]);
  });

  it("with weight=1 finds optimal path", () => {
    const grid = createEmptyGrid(5, 5);
    setCell(grid, 0, 0, "start");
    setCell(grid, 4, 4, "end");

    const result = weightedAStar(grid, [0, 0], [4, 4], 1.0);

    expect(result.path.length).toBe(9);
  });

  it("returns empty path when no route exists", () => {
    const grid = createEmptyGrid(5, 5);
    setCell(grid, 0, 0, "start");
    setCell(grid, 4, 4, "end");

    setCell(grid, 0, 1, "wall");
    setCell(grid, 1, 0, "wall");
    setCell(grid, 1, 1, "wall");

    const result = weightedAStar(grid, [0, 0], [4, 4], 1.5);

    expect(result.path).toEqual([]);
  });

  it("navigates around walls", () => {
    const grid = createEmptyGrid(5, 5);
    setCell(grid, 0, 0, "start");
    setCell(grid, 0, 4, "end");

    setCell(grid, 0, 2, "wall");
    setCell(grid, 1, 2, "wall");
    setCell(grid, 2, 2, "wall");

    const result = weightedAStar(grid, [0, 0], [0, 4], 1.5);

    expect(result.path.length).toBeGreaterThan(0);
    expect(result.path[result.path.length - 1]).toEqual([0, 4]);
  });

  it("handles adjacent start and end", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 0, 0, "start");
    setCell(grid, 0, 1, "end");

    const result = weightedAStar(grid, [0, 0], [0, 1], 1.5);

    expect(result.path).toEqual([
      [0, 0],
      [0, 1],
    ]);
  });

  it("handles start equal to end", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 1, 1, "start");

    const result = weightedAStar(grid, [1, 1], [1, 1], 1.5);

    expect(result.path.length).toBe(1);
    expect(result.path[0]).toEqual([1, 1]);
  });

  it("explores fewer nodes with higher weight", () => {
    const grid = createEmptyGrid(10, 10);
    setCell(grid, 0, 0, "start");
    setCell(grid, 9, 9, "end");

    const lowWeightResult = weightedAStar(grid, [0, 0], [9, 9], 1.0);
    const highWeightResult = weightedAStar(grid, [0, 0], [9, 9], 3.0);

    expect(highWeightResult.visited.length).toBeLessThanOrEqual(lowWeightResult.visited.length);
  });

  it("records the weight used", () => {
    const grid = createEmptyGrid(3, 3);
    setCell(grid, 0, 0, "start");
    setCell(grid, 2, 2, "end");

    const result = weightedAStar(grid, [0, 0], [2, 2], 2.0);

    expect(result.weight).toBe(2.0);
  });
});

File: src/algorithms/pathfinding/heuristic-search/weighted-a-star/__tests__/WeightedAStar_test.java
// javac WeightedAStar.java WeightedAStar_test.java && java -ea WeightedAStar_test
public class WeightedAStar_test {

    static int[][] makeEmptyGrid(int rows, int cols) {
        return new int[rows][cols];
    }

    static void setWall(int[][] grid, int row, int col) {
        grid[row][col] = 1;
    }

    public static void main(String[] args) {
        // Test: finds path on empty grid
        {
            int[][] grid = makeEmptyGrid(5, 5);
            int[][] path = WeightedAStar.weightedAStar(grid, new int[]{0, 0}, new int[]{4, 4}, 1.5);
            assert path.length > 0 : "Expected non-empty path";
            assert path[0][0] == 0 && path[0][1] == 0 : "Path should start at [0,0]";
            assert path[path.length-1][0] == 4 && path[path.length-1][1] == 4 : "Path should end at [4,4]";
        }

        // Test: with weight 1.0 finds optimal path
        {
            int[][] grid = makeEmptyGrid(5, 5);
            int[][] path = WeightedAStar.weightedAStar(grid, new int[]{0, 0}, new int[]{4, 4}, 1.0);
            assert path.length == 9 : "Expected path length 9, got " + path.length;
        }

        // Test: returns empty path when no route
        {
            int[][] grid = makeEmptyGrid(5, 5);
            setWall(grid, 0, 1);
            setWall(grid, 1, 0);
            setWall(grid, 1, 1);
            int[][] path = WeightedAStar.weightedAStar(grid, new int[]{0, 0}, new int[]{4, 4}, 1.5);
            assert path.length == 0 : "Expected empty path";
        }

        // Test: handles adjacent start and end
        {
            int[][] grid = makeEmptyGrid(3, 3);
            int[][] path = WeightedAStar.weightedAStar(grid, new int[]{0, 0}, new int[]{0, 1}, 1.5);
            assert path.length == 2 : "Expected path length 2";
        }

        // Test: handles start equal to end
        {
            int[][] grid = makeEmptyGrid(3, 3);
            int[][] path = WeightedAStar.weightedAStar(grid, new int[]{1, 1}, new int[]{1, 1}, 1.5);
            assert path.length == 1 : "Expected path length 1";
        }

        System.out.println("All tests passed!");
    }
}
